drivers/wireless/ieee80211/bcm43xxx/bcmf_sdio.h

```diff
@@ -110,13 +111,22 @@ struct bcmf_sdio_dev_s
 
 /* Structure used to manage SDIO frames */
 
+struct __attribute__((packed)) bcmf_dma_buf
+{
+  uint8_t align_padding[CONFIG_IEEE80211_BROADCOM_SDIO_DMA_BUF_ALIGNMENT -
```

Comment:
why not attach aligned_data to the char buffer directly?

Reply:
I tried to use __attribute__((aligned(16))) attached directly to "uint8_t remaining_aligned_buf", however the alignment did not work inside of struct __attribute__((packed)). (first_word[] and remaining_aligned_buf[] parts are required to be adjacent). Thus I've added align_padding[] before first_word[] and attached __attribute__((aligned(16))) to the whole structure.
One more alternative approach would be to use a monolitic data array and use offsets to point to imaginary first_word[] and remaining_aligned_buf[] parts. Is the alternative preferable?

Reply:
but bcmf_sdio_frame don't have attribute((packed)), which struct you reference has this attribute?

Reply:
I'm referencing bcmf_dma_buf struct I've introduced. first_word[] and remaining_aligned_buf[] parts should not have a gap between each other as they form the whole bcmf frame. Thus I've added "attribute((packed))" to the struct to prevent a gap inside of the struct (now I've wrapped it into begin_packed_struct / end_packed_struct for compiler independence).
However, the whole bcmf_dma_buf struct should be aligned to the address boundary of 16.
As the struct is packed inside and aligned to 16 outside, then remaining_aligned_buf[] is also aligned to the address boundary of 16, that is the purpose as it's used in DMA transfer.
There are two conditions at the same time:
- remaining_aligned_buf[] should be aligned;
- remaining_aligned_buf[] should follow first_word[] w/o a gap.
The main purpose of first_word[] is to receive the length (the first phase) of the remaining frame to be read (the second phase). DMA receive transfer can not be invoked w/o knowing the length of the frame first, thus it can not be a single transfer.

Reply:
It's more simple to attach aligned_data(16) to the data field directly:
```
struct bcmf_sdio_frame
{
  struct bcmf_frame_s header;
  bool                tx;
  dq_entry_t          list_entry;
  uint8_t             data[HEADER_SIZE + MAX_NETDEV_PKTSIZE +
                           CONFIG_NET_GUARDSIZE] aligned_data(16);
};
```
And ld.script may don't need change too. There are many examples in #4238.

Reply:
> It's more simple to attach aligned_data(16) to the data field directly:
In your example I need data[] buffer to be aligned to the boundary of 16 at the offset of 4 (not at the offset of 0). Please, check my previous message.

And as I previously wrote, one more alternative approach would be to use a monolitic data array (like in your example) and use offsets to point to imaginary first_word[] and remaining_aligned_buf[] parts. Is the alternative preferable?

Reply:
In my opinion, the approach by using bcmf_dma_buf struct is better as it's self-documented and gives a better understanding compared to a monolithic buffer with offsets in the programming code.

Reply:
Ok, after reviewing your change more carefully, you need read the first 4bytes and then the rest packet:

1. the first word align on 12bytes
2. then the rest packet align on 16bytes

So, you declare the buffer is 16bytes alignment and skip the 12bytes manually. But, why not directly align the buffer 12bytes directly liket this:
```
uint8_t             data[HEADER_SIZE + MAX_NETDEV_PKTSIZE +
                           CONFIG_NET_GUARDSIZE] aligned_data(CONFIG_IEEE80211_BROADCOM_SDIO_DMA_BUF_ALIGNMENT - 4);
```

Reply:
> So, you declare the buffer is 16bytes alignment and skip the 12bytes manually. But, why not directly align the buffer 12bytes directly liket this:
> 
> ```
> uint8_t data[HEADER_SIZE + MAX_NETDEV_PKTSIZE + CONFIG_NET_GUARDSIZE] 
>    aligned_data(CONFIG_IEEE80211_BROADCOM_SDIO_DMA_BUF_ALIGNMENT - 4);
> ```

Any alignment must be power of 2 (it's the gcc standard requirement):

```
wireless/ieee80211/bcm43xxx/bcmf_sdio.h:130:3: error: requested alignment is not a positive power of 2
   uint8_t data2[32] aligned_data(12);
   ^~~~~~~
```

Reply:
Ok, but from your change, align_padding and remaining_aligned_buf is never referenced. How about we change to:
```
struct bcmf_sdio_frame
{
  struct bcmf_frame_s header;
  bool                tx;
  dq_entry_t          list_entry;
  uint8_t             pad[CONFIG_IEEE80211_BROADCOM_SDIO_DMA_BUF_ALIGNMENT - FIRST_WORD_SIZE] aligned_data(CONFIG_IEEE80211_BROADCOM_SDIO_DMA_BUF_ALIGNMENT);
  uint8_t             data[HEADER_SIZE + MAX_NETDEV_PKTSIZE + CONFIG_NET_GUARDSIZE];
};
```
Which could reduce the rest change.

Reply:
How about this suggestion, @a-lunev ?

Reply:
Your suggestion is good. Thanks. Just now I've applied it and tested it on Photon board.